src/Codec/Compression/LZ4/Conduit.hsc

```diff
@@ -469,41 +503,44 @@ compressWithOutBufferSize bufferSize =
 
     outBuf <- liftIO $ mallocForeignPtrBytes (fromIntegral outBufferSize)
     let withOutBuf f = liftIO $ withForeignPtr outBuf f
+
     let yieldOutBuf outBufLen = do
           outBs <- withOutBuf $ \buf -> packCStringLen (buf, fromIntegral outBufLen)
-          yield outBs
+          yield (Chunk outBs)
 
-    headerSize <- withOutBuf $ \buf -> lz4fCompressBegin ctx prefs buf outBufferSize
+    let writeHeader = withOutBuf $ \buf -> lz4fCompressBegin ctx prefs buf outBufferSize
 
     let writeFooterAndYield remainingCapacity = do
-          let offset = fromIntegral $ outBufferSize - remainingCapacity
-          footerWritten <- withOutBuf $ \buf -> lz4fCompressEnd ctx (buf `plusPtr` offset) remainingCapacity
-          let outBufLen = outBufferSize - remainingCapacity + footerWritten
-          yieldOutBuf outBufLen
+          let go capacity = do
+                let offset = fromIntegral $ outBufferSize - capacity
+                footerWritten <- withOutBuf $ \buf -> lz4fCompressEnd ctx (buf `plusPtr` offset) capacity
+                let outBufLen = outBufferSize - capacity + footerWritten
+                yieldOutBuf outBufLen
+                yield Flush -- denote the end of a frame.
+
+          -- Passing srcSize==0 provides bound for LZ4F_compressEnd(),
+          -- see docs of LZ4F_compressBound() for that.
+          footerSize <- liftIO $ lz4fCompressBound 0 prefs
+
+          if remainingCapacity >= footerSize
+            then do
+              go remainingCapacity
+            else do
+              -- Footer doesn't fit: Yield buffer, put footer into now-free buffer
+              yieldOutBuf (outBufferSize - remainingCapacity)
+              go outBufferSize
 
     let loop remainingCapacity = do
```

Comment:
I'm not super convinced this actually does what we hope it does. We're reusing the same compression context across frames. From what I understand in reading about LZ77, wouldn't you then be able to encounter tokens, whose uncompressed versions occurred in an earlier frame?

I believe the safest alternative is to implement `compressWithOutBufferSizeMultiFrame` by calling/restarting `compressWithOutBufferSize` on each `Flush`.

Reply:
The struct for bookkeeping the compression context is being reused, not the actual compression context itself.  From the LZ4 manual:
> A successful call to LZ4F_compressEnd() makes `cctx` available again for another compression task.

I've verified this by reading the source for `LZ4F_compressBegin` and `LZ4F_compressEnd`. Will add a comment for this.